gnc/sub8_perception/src/sub8_pcl/cv_utils.cc

```diff
@@ -1,12 +1,270 @@
 #include <sub8_pcl/cv_tools.hpp>
 
 namespace sub {
+
 cv::Point contour_centroid(Contour& contour) {
   cv::Moments m = cv::moments(contour, true);
   cv::Point center(m.m10 / m.m00, m.m01 / m.m00);
   return center;
 }
 
+
+cv::MatND smooth_histogram(const cv::MatND &histogram, size_t filter_kernel_size, float sigma){
+	cv::MatND hist = histogram.clone();
+	std::vector<float> gauss_kernel = generate_gaussian_kernel_1D(filter_kernel_size, sigma);
+	size_t histSize = hist.total();
+	size_t offset = (filter_kernel_size - 1) / 2;
+	for (size_t i = offset; i < histSize-offset; i++ )   // Convolve histogram values with gaussian kernel
+	{
+		int sum = 0;
+		int kernel_idx = 0;
+		for (int j = i - offset; j <= i + offset; j++){
+			sum += (hist.at<float>(j) * gauss_kernel[kernel_idx++]);
+	    }
+	    hist.at<float>(i) = sum;
+	}
+	for (int i = 0; i < offset; ++i)  // Pad filtered result with zeroes
+	{
+		hist.at<float>(i) = 0;
+		hist.at<float>(histSize - 1 - i) = 0;
+	}
+	return hist;
+}
+
+
+std::vector<float> generate_gaussian_kernel_1D(size_t kernel_size, float sigma){
+	std::vector<float> kernel;
+	int middle_index = (kernel_size - 1) / 2;
+	int first_discrete_sample_x = -(middle_index);
+	for (int i = first_discrete_sample_x; i <= 0; i++)
+	{
+		float power = -0.5 * (float(i)/sigma) * (float(i)/sigma);
+		kernel.push_back( exp(power) );				// From definition of Standard Normal Distribution
+	}
+	for(int i = 1; i <= middle_index; i++){			// Kernel is symmetric
+		kernel.push_back(kernel[middle_index - i]);
+	}
+	// Normalize kernel (sum of values should equal 1.0)
+	float sum = 0;
+	for (size_t i = 0; i < kernel_size; i++){ sum += kernel[i];	}
+	for (size_t i = 0; i < kernel_size; i++){ kernel[i] /= sum;	}
+	return kernel;
+}
+
+
+std::vector<cv::Point> find_local_maxima(const cv::MatND &histogram, float thresh_multiplier){
+	std::vector<cv::Point> local_maxima, threshed_local_maxima;
+	float global_maximum = 0;
+
+	// Locate local maxima and find global maximum
+	for(size_t idx = 1; idx < histogram.total() - 1; idx++){
+		float current_value = histogram.at<float>(idx);
+	    if((histogram.at<float>(idx-1) < current_value) && (histogram.at<float>(idx+1) <= current_value)){
```

Comment:
No issue since speed is not  concern, but you could save some cycles if you cache previously accessed indices. I don't think opencv mats are structured in a way that lets the compiler figure this out.